### PR TITLE
add series axis

### LINKED ISSUE #33: add series axis

### PR DESCRIPTION
**Pull request by umarmohammed**
closes #33

### CODE REVIEW

File: projects/ngx-d3timeline/src/lib/axis/serie-axis-view-model.ts [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+import { View } from '../view/view';
+
+export class SeriesAxisViewModel {
+  private readonly bandscale: ScaleBand<string>;
+
+  constructor(data: TimelineEvent[], view: View) {
+    this.bandscale = this.createBandscale(data, view);
+  }
+
+  get tickInfo(): { tick: string; tickTransform: string }[] {
```
Comment (wflemingnz-retired) [inline projects/ngx-d3timeline/src/lib/axis/serie-axis-view-model.ts]:
```suggestion
  get tickInfos(): { label: string; transform: string }[] {
```

File: projects/ngx-d3timeline/src/lib/axis/serie-axis-view-model.ts [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+
+  get rangeLimit() {
+    return this.bandscale.range()[1];
+  }
+
+  get bandwidth() {
+    return this.bandscale.bandwidth();
+  }
+
+  private createBandscale(data: TimelineEvent[], view: View) {
```
Comment (wflemingnz-retired) [inline projects/ngx-d3timeline/src/lib/axis/serie-axis-view-model.ts]: configureScaleBand

File: projects/ngx-d3timeline/src/lib/axis/series-axis-view-model.spec.ts [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+    height: 400,
+    width: 300,
+    margin: 50
+  };
+
+  beforeEach(() => {
+    vm = new SeriesAxisViewModel(timelineEventData, view);
+  });
+
+  it('should should return correct tickInfo', () => {
```
Comment (wflemingnz-retired) [inline projects/ngx-d3timeline/src/lib/axis/series-axis-view-model.spec.ts]: remove double should

File: projects/ngx-d3timeline/src/lib/axis/serie-axis-view-model.ts [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+    return scaleBand()
+      .domain([...new Set(data.map(d => d.series))])
+      .range([0, view.width - view.margin]);
```
Comment (wflemingnz-retired) [inline projects/ngx-d3timeline/src/lib/axis/serie-axis-view-model.ts]: move left/right extreme calculation logic to new TimelineView class also with height, width, margin properties

```suggestion
      .range([timelineView.bounds.left, timelineView.bounds.right]);
```

File: projects/ngx-d3timeline/src/lib/axis/series-axis-view-model.spec.ts [hunk 5]
```diff
@@ -0,0 +1,24 @@
...
+    expect(vm.tickInfo).toMatchSnapshot();
+  });
+
+  it('should return correct rangeLimit', () => {
+    expect(vm.rangeLimit).toBe(view.width - view.margin);
```
Comment (wflemingnz-retired) [inline projects/ngx-d3timeline/src/lib/axis/series-axis-view-model.spec.ts]: expect(vm.rangeLimit).toBe(view.bounds.maxX);

File: projects/ngx-d3timeline/src/lib/axis/series-axis.component.ts [hunk 6]
```diff
@@ -0,0 +1,31 @@
+import { Component, Input } from '@angular/core';
+import { SeriesAxisViewModel } from './serie-axis-view-model';
+
+@Component({
+  selector: '[ngx-d3timeline-series-axis]',
+  template: `
+    <svg:g class="series-axis-group">
```
Comment (wflemingnz-retired) [inline projects/ngx-d3timeline/src/lib/axis/series-axis.component.ts]: rename series to resources

ie: resources-axis-group

File: projects/ngx-d3timeline/src/lib/axis/series-axis.component.ts [hunk 7]
```diff
@@ -0,0 +1,31 @@
...
+      <text
+        *ngFor="let tickInfo of vm?.tickInfo"
```
Comment (wflemingnz-retired) [inline projects/ngx-d3timeline/src/lib/axis/series-axis.component.ts]: add ngIf above and remove ?. from here

File: projects/ngx-d3timeline/src/lib/axis/series-axis.component.ts [hunk 8]
```diff
@@ -0,0 +1,31 @@
...
+        [attr.transform]="tickInfo.tickTransform"
+        dy="-2"
+      >
+        {{ tickInfo.tick }}
+      </text>
+      <svg:line class="series-axis-line" [attr.x2]="vm?.rangeLimit"></svg:line>
```
Comment (wflemingnz-retired) [inline projects/ngx-d3timeline/src/lib/axis/series-axis.component.ts]: don't need ?.

File: projects/ngx-d3timeline/src/lib/ngx-d3timeline.component.ts [hunk 9]
```diff
@@ -11,7 +11,12 @@ import { ScaleService } from './scale.service';
       [attr.height]="viewService.view.height"
       class="ngx-d3timeline"
     >
-      <g class="root-group" [attr.transform]="viewService.rootTransform"></g>
+      <g [attr.transform]="viewService.rootTransform">
```
Comment (wflemingnz-retired) [inline projects/ngx-d3timeline/src/lib/ngx-d3timeline.component.ts:14]: not needed due to viewDimensions.bounds

File: projects/ngx-d3timeline/src/lib/scale.service.ts [hunk 10]
```diff
@@ -1,11 +1,20 @@
 import { Injectable } from '@angular/core';
 import { BehaviorSubject } from 'rxjs';
+import { map } from 'rxjs/operators';
 import { TimelineEvent } from './timeline-event';
+import { SeriesAxisViewModel } from './axis/serie-axis-view-model';
+import { ViewService } from './view/view.service';
 
 @Injectable({ providedIn: 'root' })
 export class ScaleService {
```
Comment (wflemingnz-retired) [inline projects/ngx-d3timeline/src/lib/scale.service.ts]: AxisService

File: projects/ngx-d3timeline/src/lib/timeline-event-data.ts [hunk 11]
```diff
@@ -0,0 +1,70 @@
+import { TimelineEvent } from './timeline-event';
+
+export const timelineEventData: TimelineEvent[] = [
```
Comment (wflemingnz-retired) [inline projects/ngx-d3timeline/src/lib/timeline-event-data.ts]: mockData

File: projects/ngx-d3timeline/src/lib/axis/serie-axis-view-model.ts [hunk 12]
```diff
@@ -0,0 +1,37 @@
...
+  }
+
+  private tickTransform(tick: string) {
+    return `translate(${this.bandscale(tick) +
+      this.bandscale.bandwidth() / 2}, 0)`;
```
Comment (wflemingnz-retired) [inline projects/ngx-d3timeline/src/lib/axis/serie-axis-view-model.ts]:
```suggestion
      return `translate(${this.getBandMidPoint(tick)}, 0)`;
      
      getBandMidPoint(tick)
      {
          return this.bandscale(tick) + this.bandscale.bandwidth();
      }
      
```

File: projects/ngx-d3timeline/src/lib/axis/serie-axis-view-model.ts [hunk 13]
```diff
@@ -0,0 +1,37 @@
+import { ScaleBand, scaleBand } from 'd3-scale';
+import { TimelineEvent } from '../timeline-event';
+import { View } from '../view/view';
+
+export class SeriesAxisViewModel {
```
Comment (wflemingnz-retired) [inline projects/ngx-d3timeline/src/lib/axis/serie-axis-view-model.ts:5]:
```suggestion
export class ResourcesAxisViewModel {
```

**Review by wflemingnz-retired — CHANGES_REQUESTED** [review]
made a few more minor changes, let me know if any needs dicsussing

File: projects/ngx-d3timeline/src/lib/axis/serie-axis-view-model.ts (hunk 13 above)
Comment (umarmohammed) [reply projects/ngx-d3timeline/src/lib/axis/serie-axis-view-model.ts:5]: will be done in #41

File: projects/ngx-d3timeline/src/lib/axis/serie-axis-view-model.ts (hunk 2 above)
Comment (umarmohammed) [reply projects/ngx-d3timeline/src/lib/axis/serie-axis-view-model.ts]: done

File: projects/ngx-d3timeline/src/lib/axis/serie-axis-view-model.ts (hunk 4 above)
Comment (umarmohammed) [reply projects/ngx-d3timeline/src/lib/axis/serie-axis-view-model.ts]: done

File: projects/ngx-d3timeline/src/lib/axis/serie-axis-view-model.ts (hunk 12 above)
Comment (umarmohammed) [reply projects/ngx-d3timeline/src/lib/axis/serie-axis-view-model.ts]: done

File: projects/ngx-d3timeline/src/lib/axis/series-axis-view-model.spec.ts (hunk 3 above)
Comment (umarmohammed) [reply projects/ngx-d3timeline/src/lib/axis/series-axis-view-model.spec.ts]: done

File: projects/ngx-d3timeline/src/lib/axis/series-axis-view-model.spec.ts (hunk 5 above)
Comment (umarmohammed) [reply projects/ngx-d3timeline/src/lib/axis/series-axis-view-model.spec.ts]: done

File: projects/ngx-d3timeline/src/lib/axis/series-axis.component.ts (hunk 6 above)
Comment (umarmohammed) [reply projects/ngx-d3timeline/src/lib/axis/series-axis.component.ts]: will be addressed in #41

File: projects/ngx-d3timeline/src/lib/axis/series-axis.component.ts (hunk 8 above)
Comment (umarmohammed) [reply projects/ngx-d3timeline/src/lib/axis/series-axis.component.ts]: done

File: projects/ngx-d3timeline/src/lib/ngx-d3timeline.component.ts (hunk 9 above)
Comment (umarmohammed) [reply projects/ngx-d3timeline/src/lib/ngx-d3timeline.component.ts:14]: not so straight forward have created a new issue #44

File: projects/ngx-d3timeline/src/lib/scale.service.ts (hunk 10 above)
Comment (umarmohammed) [reply projects/ngx-d3timeline/src/lib/scale.service.ts]: done

File: projects/ngx-d3timeline/src/lib/timeline-event-data.ts (hunk 11 above)
Comment (umarmohammed) [reply projects/ngx-d3timeline/src/lib/timeline-event-data.ts]: done

**Pull request merged**